futures-async-macro/src/lib.rs

```diff
@@ -247,6 +249,250 @@ pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
+    let attribute = quote! { #[async_stream #attribute ] };
+    let attribute = SynomBuffer::new(attribute.into());
+    let (_, attribute) = Attribute::parse_outer(attribute.begin())
+        .expect("failed to parse attribute arguments");
+    let attribute = attribute.meta_item()
+        .expect("failed to parse attribute arguments");
+
+    let mut boxed = false;
+    let mut item_ty = None;
+
+    match attribute {
+        MetaItem::Term(_) => (),
+        MetaItem::List(list) => {
+            for item in list.nested.items() {
+                if let NestedMetaItem::MetaItem(ref item) = *item {
+                    if let MetaItem::Term(ref term) = *item {
+                        if term == "boxed" {
+                            boxed = true;
+                        }
+                    }
+                    if let MetaItem::NameValue(ref item) = *item {
+                        if item.ident == "item" {
+                            let mut s = item.lit.to_string();
+                            assert_eq!(s.remove(0), '"');
+                            assert_eq!(s.pop(), Some('"'));
+                            let term = Term::intern(&s);
+                            let span = item.lit.span.clone();
+                            item_ty = Some(Path::from(Ident::new(term, span)));
+                        }
+                    }
+                }
+            }
+        }
+        MetaItem::NameValue(_) => {
+            panic!("#[async_stream] should not be invoked as name value attribute");
+        }
+    }
+
+    let boxed = boxed;
+    let item_ty = item_ty.expect("#[async_stream] requires item type to be specified");
+
+    // Parse our item, expecting a function. This function may be an actual
+    // top-level function or it could be a method (typically dictated by the
+    // arguments). We then extract everything we'd like to use.
+    let Item { attrs, node } = syn::parse(function)
+        .expect("failed to parse tokens as a function");
+    let ItemFn {
+        ident,
+        vis,
+        unsafety,
+        constness,
+        abi,
+        block,
+        decl,
+        ..
+    } = match node {
+        ItemKind::Fn(item) => item,
+        _ => panic!("#[async] can only be applied to functions"),
+    };
+    let FnDecl {
+        inputs,
+        output,
+        variadic,
+        generics,
+        fn_token,
+        ..
+    } = { *decl };
+    let where_clause = &generics.where_clause;
+    assert!(!variadic, "variadic functions cannot be async");
+    let (output, rarrow_token) = match output {
+        FunctionRetTy::Ty(t, rarrow_token) => (t, rarrow_token),
+        FunctionRetTy::Default => {
+            (TyTup {
+                tys: Default::default(),
+                lone_comma: Default::default(),
+                paren_token: Default::default(),
+            }.into(), Default::default())
+        }
+    };
+
+    // We've got to get a bit creative with our handling of arguments. For a
```

Comment:
Could this avoid duplication with `#[async]` and be refactored to a common implementation?